cwe/cwe_test.go

```diff
@@ -0,0 +1,90 @@
+package cwe
+
+import (
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/http/httptest"
+	"os"
+	"path/filepath"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestUpdate(t *testing.T) {
+	testCases := []struct {
+		name                   string
+		inputZipFile           string
+		expectedOuptutXMLFile  string
+		expectedOutputJSONFile string
+		expectedError          string
+		cweServerUrl           string
+	}{
+		{
+			name:                   "happy path",
+			inputZipFile:           "goldens/good-small-cwe.xml.zip",
+			expectedOuptutXMLFile:  "goldens/good-small-cwe.xml",
+			expectedOutputJSONFile: "goldens/good-small-cwe.json",
+		},
+		{
+			name:          "sad path, corrupt xml file in zip",
+			inputZipFile:  "goldens/corrupt.xml.zip",
+			expectedError: "XML syntax error",
+		},
+		{
+			name:          "sad path, invalid zip file",
+			inputZipFile:  "goldens/bad.xml.zip",
+			expectedError: "not a valid zip file",
+		},
+		{
+			name:          "sad path, too many files in archive",
+			inputZipFile:  "goldens/toomanyfiles.xml.zip",
+			expectedError: "too many files in archive",
+		},
+		{
+			name:          "sad path, unreachable CWE service",
+			expectedError: "failed to fetch cwe data",
+			cweServerUrl:  "http://foo/bar/baz",
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			var cweURL string
+			if tc.cweServerUrl != "" {
+				cweURL = tc.cweServerUrl
+			} else {
+				ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+					b, _ := ioutil.ReadFile(tc.inputZipFile)
+					_, _ = io.WriteString(w, string(b))
+				}))
+				cweURL = ts.URL
+				defer func() {
+					ts.Close()
+				}()
+			}
+
+			dir, _ := ioutil.TempDir("", "TestUpdate-*")
```

Comment:
nit: We can ignore this error, but it will introduce another error at `c.Update()` when it fails to create the temp dir. In that case, it may obscure the root cause.

Reply:
I think it's fine as is, it would be a failure either way.

Reply:
Yes, it's no big deal, but I mean it would take a long time to figure out the root cause. For example,

```
a, _ := func1()
b, _ := func2()
c, _ := func3()
err := func4(a, b, c) // <- this might fail due to the error of func1/2/3
assert.NoError(t, err)
```

In this case, as you said, it would be a failure either way at func4(). But it is difficult to know if the error comes from func1, func2, func3, or func4. If func1() creates an instance (in this example, `a`) representing a network socket, func4() may cause a panic of nil pointer when the network connection can't be established and `a` is nil. The error from func1() should describe the detail like `connect to address 127.0.0.1: Connection refused`. When we get a nil pointer panic, we can't know that the root cause is a network problem quickly. If we have assertions for each error, I think it would be helpful for debugging.
In addition, it might cause unexpected behavior. If `ioutil.TempDir` fails and returns an empty string, the files are created under the current directory and the current directory will be removed after this test because it has `defer`.

```
			defer func() {
				_ = os.RemoveAll(dir)
			}()
```

I don't strongly recommend adding the assertion as I said `nit`, but for the reasons above, I basically prefer adding an error check for each err if I don't have a reason to ignore the error.